# **Personalized Experience with Face Recognition**

## <center>By : Group 8 - Yu-Fang(Unice) Liao, Han Nga Chan, Yumeng Tang</center>

## MOTIVATION & PROBLEM STATEMENT

Retaining existing customers are paramount objectives within the retail industry. Achieving this relies heavily on providing a seamless customer experience to optimize service efficiency. Despite the widespread popularity of online shopping, approximately 73% of customers still prefer the in-store experience. Research underscores that customers seek minimal time spent on the shopping floor coupled with personalized service, a dynamic that cultivates loyalty. Offline shops face hurdles in meeting customer needs due to limited access to relevant information, leading to delays in providing product details and verifying memberships at checkout, as highlighted by IBM. Swiftly gathering customer information upon store entry is crucial to address these challenges. Implementing facial recognition technology at the entrance, with customer consent, enables staff to quickly recognize preferences, offer timely product information, and streamline checkout by eliminating the need for customers to repeat membership details. Leveraging facial recognition empowers offline stores to greatly enhance the customer experience, boost retention rates, and ultimately drive sales and profitability.


## DATASET OVERVIEW

We used Pins Face Recognition datasets (https://www.kaggle.com/datasets/hereisburak/pins-face-recognition/data) from Kaggle and online image scraps for retail store facial recognition. The dataset includes 105 celebrities with around 38,500 high-quality images from Pinterest, covering various demographics and expressions. Among them, 55 are actresses and 50 are actors. This diverse dataset is crucial for training accurate facial detection models in real-world retail scenarios, enhancing personalized in-store experiences for customers.

# Install the package & import the libraries

In [ ]:
pip install bing-image-downloader

Note: you may need to restart the kernel to use updated packages.


In [ ]:
import os
import shutil
from bing_image_downloader import downloader

# Web Scraping

In order to increase the data for each class, we use a web crawler to collect 200 photos for each class (each celebrity) and store them in the corresponding folder.

In [ ]:
# Web scraping - Version 1: collect 200 photos
source_directory = "img/105_classes_pins_dataset/"
output_directory = "img/new_web_scraping/0421/"
print("Source directory:", source_directory)
print("Output directory:", output_directory)

os.makedirs(output_directory, exist_ok=True)
print("Output directory ensured.")

for item in os.listdir(source_directory):
    item_path = os.path.join(source_directory, item)
    if os.path.isdir(item_path):
        name = item.split("pins_")[-1]
        folder_path = os.path.join(output_directory, name)
        os.makedirs(folder_path, exist_ok=True)
        print(f"Processing and downloading images for {name}.")

        try:
            downloader.download(name, limit=200, output_dir=folder_path, adult_filter_off=True, force_replace=False, timeout=60, verbose=True)
            # Adjust for unwanted subdirectories created by the downloader
            unwanted_subdirectory = os.path.join(folder_path, name)
            if os.path.exists(unwanted_subdirectory):
                for file in os.listdir(unwanted_subdirectory):
                    shutil.move(os.path.join(unwanted_subdirectory, file), folder_path)
                os.rmdir(unwanted_subdirectory)
            print(f"Images successfully downloaded for {name}.")
        except Exception as e:
            print(f"Error during download for {name}: {e}")
    else:
        print(f"Skipped: {item} (Not a directory)")

Source directory: img/105_classes_pins_dataset/
Output directory: img/new_web_scraping/0421/
Output directory ensured.
Processing and downloading images for Irina Shayk.
[%] Downloading Images to /projectnb/ba865/projects/Group8_B1/img/new_web_scraping/0421/Irina Shayk/Irina Shayk


[!!]Indexing page: 1

[%] Indexed 35 Images on Page 1.


[%] Downloading Image #1 from http://www.hawtcelebs.com/wp-content/uploads/2012/05/IRINA-SHAYK-at-Killing-Them-Softly-Premiere-at-Cannes-Film-Festival-4.jpg
[%] File Downloaded !

[%] Downloading Image #2 from http://www.hawtcelebs.com/wp-content/uploads/2012/02/Irina-Shayk-at-Elton-John-AIDS-Foundation-Academy-Awards-Viewing-Party-in-Beverly-Hills-6.jpg
[%] File Downloaded !

[%] Downloading Image #3 from https://4.bp.blogspot.com/-rpfSUXwLJME/UhyafQ1QWhI/AAAAAAAAJis/1C4vjaB-8rc/s1600/Irina+Shayk+Photo+2013+09.jpg
[%] File Downloaded !

[%] Downloading Image #4 from https://www.sheknows.com/wp-content/uploads/2023/06/irina-shayk.jpg
[%] File Download

However, due to an interruption during the use of SCC, the web crawling data is incomplete. Therefore, we use this code to check the number of existing pictures in each folder. If the number is less than 200, we continue the web crawling.

In [ ]:
# If the number of the web scraping pictures in the folder is less than 200, then continue download new pictures.
source_directory = "img/105_classes_pins_dataset/"
output_directory = "img/new_web_scraping/0421/"
print("Source directory:", source_directory)
print("Output directory:", output_directory)

os.makedirs(output_directory, exist_ok=True)
print("Output directory ensured.")

for item in os.listdir(source_directory):
    item_path = os.path.join(source_directory, item)
    if os.path.isdir(item_path):
        name = item.split("pins_")[-1]
        folder_path = os.path.join(output_directory, name)
        os.makedirs(folder_path, exist_ok=True)
        print(f"Checking and processing images for {name}.")

        # Count existing images
        existing_images_count = len([f for f in os.listdir(folder_path) if os.path.isfile(os.path.join(folder_path, f))])
        print(f"Found {existing_images_count} existing images in {name}.")

        if existing_images_count < 200:
            needed_images = 200 - existing_images_count
            print(f"Downloading additional {needed_images} images to complete 200.")

            try:
                downloader.download(name, limit=needed_images, output_dir=folder_path, adult_filter_off=True, force_replace=False, timeout=60, verbose=True)
                # Adjust for unwanted subdirectories created by the downloader
                unwanted_subdirectory = os.path.join(folder_path, name)
                if os.path.exists(unwanted_subdirectory):
                    for file in os.listdir(unwanted_subdirectory):
                        shutil.move(os.path.join(unwanted_subdirectory, file), folder_path)
                    os.rmdir(unwanted_subdirectory)
                print(f"Images successfully downloaded for {name}.")
            except Exception as e:
                print(f"Error during download for {name}: {e}")
        else:
            print(f"No additional images needed for {name}.")
    else:
        print(f"Skipped: {item} (Not a directory)")

Source directory: img/105_classes_pins_dataset/
Output directory: img/new_web_scraping/0421/
Output directory ensured.
Checking and processing images for Irina Shayk.
Found 200 existing images in Irina Shayk.
No additional images needed for Irina Shayk.
Checking and processing images for Katherine Langford.
Found 200 existing images in Katherine Langford.
No additional images needed for Katherine Langford.
Checking and processing images for Taylor Swift.
Found 200 existing images in Taylor Swift.
No additional images needed for Taylor Swift.
Checking and processing images for Jessica Barden.
Found 200 existing images in Jessica Barden.
No additional images needed for Jessica Barden.
Checking and processing images for Jimmy Fallon.
Found 200 existing images in Jimmy Fallon.
No additional images needed for Jimmy Fallon.
Checking and processing images for Richard Harmon.
Found 200 existing images in Richard Harmon.
No additional images needed for Richard Harmon.
Checking and processing im

Finally, we will merge the dataset downloaded from Kaggle (105_classes_pins_dataset) with the data obtained from web crawling, placing photos of the same celebrity into the same folder.

In [ ]:
# Combine the web scraping folder and the dataset from Kaggle
source_dir = "/projectnb/ba865/projects/Group8_B1/img/105_classes_pins_dataset"
web_scraping_dir = "/projectnb/ba865/projects/Group8_B1/img/new_web_scraping/0421"
output_dir = "/projectnb/ba865/projects/Group8_B1/img/combine105_webscraping"

os.makedirs(output_dir, exist_ok=True)
web_scraping_folders = os.listdir(web_scraping_dir)

for folder in web_scraping_folders:
    pin_folder_name = f"pins_{folder}"
    source_folder_path = os.path.join(source_dir, pin_folder_name)
    web_scraping_folder_path = os.path.join(web_scraping_dir, folder)
    combined_folder_path = os.path.join(output_dir, folder)

    if os.path.isdir(source_folder_path):
        # Create combined folder
        os.makedirs(combined_folder_path, exist_ok=True)

        # Copy files from a source to a destination
        def copy_files(src, dst):
            for file_name in os.listdir(src):
                full_file_name = os.path.join(src, file_name)
                if os.path.isfile(full_file_name):
                    shutil.copy(full_file_name, dst)

        # Copy files from both source and web scraping folders to the combined folder
        copy_files(source_folder_path, combined_folder_path)
        copy_files(web_scraping_folder_path, combined_folder_path)

        print(f"Files from {source_folder_path} and {web_scraping_folder_path} have been combined in {combined_folder_path}")
    else:
        print(f"No matching source folder found for {folder}")

Files from /projectnb/ba865/projects/Group8_B1/img/105_classes_pins_dataset/pins_Selena Gomez and /projectnb/ba865/projects/Group8_B1/img/new_web_scraping/0421/Selena Gomez have been combined in /projectnb/ba865/projects/Group8_B1/img/combine105_webscraping/Selena Gomez
Files from /projectnb/ba865/projects/Group8_B1/img/105_classes_pins_dataset/pins_Brenton Thwaites and /projectnb/ba865/projects/Group8_B1/img/new_web_scraping/0421/Brenton Thwaites have been combined in /projectnb/ba865/projects/Group8_B1/img/combine105_webscraping/Brenton Thwaites
Files from /projectnb/ba865/projects/Group8_B1/img/105_classes_pins_dataset/pins_alycia dabnem carey and /projectnb/ba865/projects/Group8_B1/img/new_web_scraping/0421/alycia dabnem carey have been combined in /projectnb/ba865/projects/Group8_B1/img/combine105_webscraping/alycia dabnem carey
Files from /projectnb/ba865/projects/Group8_B1/img/105_classes_pins_dataset/pins_Anthony Mackie and /projectnb/ba865/projects/Group8_B1/img/new_web_scrapi

# Facial Detection

We perform Facial Detection through Haar Cascade Classifier. It helps to detect and extract the headshots of the pictures.

In [ ]:
import cv2
import os
import numpy as np
import matplotlib.pyplot as plt
import random
from PIL import Image, ImageOps


# Define the directory containing the images
main_directory = 'combine105_webscraping'

# Dimension of images
image_width = 224
image_height = 224
skipped_photos = 0
# Load the pre-trained Haar cascade classifier for face detection
facecascade = cv2.CascadeClassifier('haarcascade_frontalface_default.xml')

# Get a list of all folders
all_folders = [folder for folder in os.listdir(main_directory) if os.path.isdir(os.path.join(main_directory, folder))]

# Iterate through selected folders and display one random image from each folder
for i, folder in enumerate(all_folders):
    folder_path = os.path.join(main_directory, folder)
    all_images = [img for img in os.listdir(folder_path) if img.lower().endswith(('.jpg', '.jpeg', '.png'))]
    for image_file in all_images:
        image_path = os.path.join(folder_path, image_file)

#     # Get the full path to the selected image
#     image_path = os.path.join(folder_path, all_images)

    # Load the image
    # Load the image in BGR color space
        img_bgr = cv2.imread(image_path, cv2.IMREAD_COLOR)

        if img_bgr is None:
            print(f"Failed to load image: {image_path}")
            continue

# Convert the color space from BGR to RGB
        imgtest = cv2.cvtColor(img_bgr, cv2.COLOR_BGR2RGB)

    # Get the faces detected in the image
        faces = facecascade.detectMultiScale(imgtest, scaleFactor=1.05, minNeighbors=7)

    # If not exactly 1 face is detected, skip this photo
        if len(faces) != 1:
            print(f'---Photo skipped---\n')
            skipped_photos += 1
            continue

        for (x_, y_, w, h) in faces:
        # draw the face detected
#         face_detect = cv2.rectangle(imgtest, (x_, y_), (x_+w, y_+h), (255, 0, 0), 2)
#         plt.imshow(face_detect)
#         plt.show()
        # Detected face region
            roi = imgtest[y_: y_ + h, x_: x_ + w]
        # Resize the detected face to 224x224
            size = (image_width, image_height)
            resized_image = cv2.resize(roi, size)
            image_array = np.array(resized_image, "uint8")

        # remove the original image
            os.remove(image_path)

        # replace the image with only the face
            im = Image.fromarray(image_array)
            im = im.convert('RGB')
            #im = cv2.cvtColor(im, cv2.COLOR_BGR2RGB)
            im.save(image_path)


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---



libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: profile 'Photoshop ICC profile': 'RGB ': RGB color space not permitted on grayscale PNG


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: extra compressed data


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: cHRM chunk does not match sRGB


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

Failed to load image: combine105_webscraping/gal gadot/Image_18.jpg
---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo

libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---



libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: known incorrect sRGB profile
libpng warning: cHRM: inconsistent chromaticities


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---



libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---



libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---



libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---



libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: known incorrect sRGB profile


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: cHRM chunk does not match sRGB


---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skipped---

---Photo skip

In [ ]:
skipped_photos

13513

About 13513 images are not processed, as more then one face is detected in the pictures.

# CONCLUSION:

In conclusion, facial recognition technology can significantly enhance the retail experience by personalizing customer experience and streamlining service operations. Based on the various advanced models’ performance results, the most optimized model is the ResNet50 model with the highest training accuracy of 0.9582 and a test accuracy of 0.8462, indicating its superior capability in handling complex image data. The combination of the Haar Cascade classifier for precise facial detection and strategic hyperparameter tuning further refined our approach, addressing key challenges such as image quality and classification accuracy. By reducing the number of classes and enriching the dataset through targeted web scraping, we successfully enhanced model reliability and performance. These advancements not only optimized customer service operations by reducing time spent during transactions but also paved the way for future enhancements in retail settings, ensuring a more efficient and enjoyable shopping experience for customers.

# NEXT STEPS & RECOMMENDATIONS:

> Based on the accuracy result and overall model performance, ResNet18 and VGG are preferable for simpler tasks with limited computational resources, while ResNet50 and ViT are more suitable for achieving higher accuracy in tasks that benefit from deeper architectures or attention mechanisms.

> Moving forward, we will raise the model accuracy by adjusting the parameters of the current model, reducing the number of classes but increasing the number of images while training, increasing the number of images for each celebrity, and also implementing different augmentation methods.

> By incorporating the model, we can implement the model to achieve real-time facial capture in retail stores, detecting real-time faces for customers from the camera and automatically popping up the membership information to reduce the check-out time and this efficient payment system allows customers to complete transactions securely and conveniently. Ultimately, this system can enhance in-store customer experience and improve retailers' operation efficiency.






# REFERENCE:

https://www.forbes.com/sites/forbestechcouncil/2024/02/06/navigating-loyalty-growth-how-retailers-can-overcome-key-challenges/?sh=5d431d1302fe.

https://www.forbes.com/sites/forbestechcouncil/2023/10/20/delivering-convenient-personalized-and-seamless-retail-experiences/?sh=51756d111e85.

https://newsroom.ibm.com/2024-01-08-IBM-Study-Widespread-Discontent-in-Retail-Experiences,-Consumers-Signal-Interest-in-AI-Driven-Shopping-Amid-Economic-Strain.

https://github.com/hardikvasa/google-images-download/pull/298

https://stackoverflow.com/questions/60370799/google-image-download-with-python-cannot-download-images

https://google-images-download.readthedocs.io/en/latest/installation.html

https://stackoverflow.com/questions/60370799/google-image-download-with-python-cannot-download-images

https://www.codemag.com/Article/2205081/Implementing-Face-Recognition-Using-Deep-Learning-and-Support-Vector-Machines

https://github.com/rcmalli/keras-vggface/tree/master/keras_vggface

https://github.com/aaronpp65/face-recognition-vggface2?tab=readme-ov-file

https://www.linkedin.com/pulse/create-face-recognition-model-using-vgg16-transfer-learning-kumar/

https://stackoverflow.com/questions/67018079/error-in-from-keras-utils-import-to-categorical

https://www.kaggle.com/code/vinayakshanawad/celebrity-face-recognition-vggface-model

https://stackoverflow.com/questions/3451111/unzipping-files-in-python

https://www.geeksforgeeks.org/how-to-move-all-files-from-one-directory-to-another-using-python/

https://www.kaggle.com/code/alkanerturan/facedetection-recognition/notebook

https://www.kaggle.com/code/saeedghamshadzai/face-rec-cnn-tf-acc-71-52

https://www.v7labs.com/blog/train-validation-test-set

https://www.kaggle.com/code/sandy1112/create-and-train-resnet50-from-scratch

https://github.com/keras-team/keras-applications/blob/master/keras_applications/resnet50.py


# Claim of the use of ChatGPT:

We have use ChatGPT in the following aspect:

*   Correcting the code for performing Haar Cascade Classifier in Facial detection.
*   Debugging the code for the ViT model training process.
*   Debugging the code for the web scraping and combining the folders.
*   Debugging the code for ResNet50 to change the input size.
*   Check how to fix the error caused by import the VGG pre-trained model
*   Adjust the classifier layer of pre-trained model to fit our image data size